### PR TITLE
[server][controller] Support meta system store schema registeration at server start time

### PR DESCRIPTION
**Pull request by shuhui-liu**
## Summary

When a server starts, if system schema initialization is enabled, it will call controller to check and register new meta system store schemas. Therefore, servers can be deployed before controllers when meta system store schema is upgraded. Controller url (for open source users who do not want to onboard d2) or d2 configs must be provided to enable system schema initialization at start time.

## How was this PR tested?

Unit test and integration test 

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/config/VeniceServerConfig.java [hunk 1]
```diff
@@ -599,6 +607,11 @@ public VeniceServerConfig(VeniceProperties serverProperties, Map<String, Map<Str
     fastClassSchemaWarmupTimeout =
         serverProperties.getLong(SERVER_SCHEMA_FAST_CLASS_WARMUP_TIMEOUT, 2 * Time.MS_PER_MINUTE);
     schemaPresenceCheckEnabled = serverProperties.getBoolean(SERVER_SCHEMA_PRESENCE_CHECK_ENABLED, true);
+    systemSchemaInitializationAtStartTimeEnabled =
+        serverProperties.getBoolean(SYSTEM_SCHEMA_INITIALIZATION_AT_START_TIME_ENABLED, false);
+    localControllerUrl = serverProperties.getString(LOCAL_CONTROLLER_URL, "");
+    localControllerD2ServiceName = serverProperties.getString(LOCAL_CONTROLLER_D2_SERVICE_NAME, "");
+    localD2ZkHost = serverProperties.getString(LOCAL_D2_ZK_HOST, "");
```
Comment (gaojieliu) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/config/VeniceServerConfig.java:614]: Why do we need both?
Can we just use d2?

File: internal/venice-common/src/main/java/com/linkedin/venice/common/VeniceSystemStoreUtils.java [hunk 2]
```diff
@@ -12,6 +14,12 @@ public class VeniceSystemStoreUtils {
   private static final String PUSH_JOB_DETAILS_STORE_NAME =
       String.format(Store.SYSTEM_STORE_FORMAT, PUSH_JOB_DETAILS_STORE);
   public static final String SEPARATOR = "_";
+  public static final UpdateStoreQueryParams DEFAULT_USER_SYSTEM_STORE_UPDATE_QUERY_PARAMS =
```
Comment (gaojieliu) [inline internal/venice-common/src/main/java/com/linkedin/venice/common/VeniceSystemStoreUtils.java:17]: This static field is the same as the one defined in the following class, so can reuse it?

File: internal/venice-common/src/main/java/com/linkedin/venice/system/store/ControllerClientBackedSystemSchemaInitializer.java [hunk 3]
```diff
@@ -97,22 +106,32 @@ public void execute() {
...
+                + discoveryResponse.getError());
+      }
+    } else {
+      String currSystemStoreCluster = discoveryResponse.getCluster();
       if (!currSystemStoreCluster.equals(clusterName)) {
         LOGGER.warn(
-            "The system store for {} already exists in cluster {}, "
-                + "which is inconsistent with the config {} which specifies that it "
-                + "should be in cluster {}. Will abort the initialization routine.",
+            "The system store for {} already exists in cluster {}, which is inconsistent with the config {} which "
```
Comment (gaojieliu) [inline internal/venice-common/src/main/java/com/linkedin/venice/system/store/ControllerClientBackedSystemSchemaInitializer.java:127]: This should be an `error` message, right?
Also I wonder whether we should return some error code for this case, so that Venice Server can choose to abort the startup when this issue happens.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceControllerConfig.java [hunk 4]
```diff
@@ -476,7 +476,7 @@ public VeniceControllerConfig(VeniceProperties props) {
     this.parentExternalSupersetSchemaGenerationEnabled =
         props.getBoolean(CONTROLLER_PARENT_EXTERNAL_SUPERSET_SCHEMA_GENERATION_ENABLED, false);
     this.systemSchemaInitializationAtStartTimeEnabled =
-        props.getBoolean(SYSTEM_SCHEMA_INITIALIZATION_AT_START_TIME_ENABLED, true);
+        props.getBoolean(SYSTEM_SCHEMA_INITIALIZATION_AT_START_TIME_ENABLED, false);
```
Comment (gaojieliu) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceControllerConfig.java:479]: Why this change?

**Review by gaojieliu — CHANGES_REQUESTED** [review]

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/config/VeniceServerConfig.java (hunk 1 above)
Comment (shuhui-liu) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/config/VeniceServerConfig.java:614]: I explained in the description that controller url is for open source users who do not want to onboard d2. I heard from Felix that some open source users do not want to run d2 https://github.com/linkedin/venice/pull/521#discussion_r1245914950

File: internal/venice-common/src/main/java/com/linkedin/venice/common/VeniceSystemStoreUtils.java (hunk 2 above)
Comment (shuhui-liu) [reply internal/venice-common/src/main/java/com/linkedin/venice/common/VeniceSystemStoreUtils.java:17]: Yes they're duplicate. I'll remove one of them.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceControllerConfig.java (hunk 4 above)
Comment (shuhui-liu) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceControllerConfig.java:479]: I made a change in this commit. When this config is enabled but neither url nor d2 config is provided, throw exception instead of aborting the initialization and continuing that startup of controller.
```
throw new VeniceException(
            "System schema initialization is enabled but neither controller url nor d2 config is provided.");
```
If the default of this config continues to be true, open source users who do not provide controller url or d2 will fail to start controllers. So I changed it to false.

File: internal/venice-common/src/main/java/com/linkedin/venice/system/store/ControllerClientBackedSystemSchemaInitializer.java (hunk 3 above)
Comment (shuhui-liu) [reply internal/venice-common/src/main/java/com/linkedin/venice/system/store/ControllerClientBackedSystemSchemaInitializer.java:127]: Yeah the system store should exist in the system schema cluster. There must be something wrong if it exists in another cluster. I'm ok with throwing an exception to abort server/controller startup when this issue happens.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceControllerConfig.java (hunk 4 above)
Comment (gaojieliu) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceControllerConfig.java:479]: I see.
Let us make sure we have this config enabled in LinkedIn env.

**Review by gaojieliu — COMMENTED** [review]
Sorry for the late follow up.
Do we have a test to validate the new server behavior when enabling system store schema registration? I didn't see it from this PR.

**Comment by shuhui-liu** [comment]
> Sorry for the late follow up. Do we have a test to validate the new server behavior when enabling system store schema registration? I didn't see it from this PR.

No problem. Since VeniceServerWrapper has the following code
```
.put(SYSTEM_SCHEMA_INITIALIZATION_AT_START_TIME_ENABLED, true)
```
, when any server starts during the integration test, ControllerClientBackedSystemSchemaInitializer will be executed. If something goes wrong during the initialization, server will fail to start and the integration test will fail. Let me know if you think that more testings need to be added.

File: services/venice-server/src/main/java/com/linkedin/venice/server/VeniceServer.java [hunk 5]
```diff
@@ -209,32 +211,64 @@ private List<AbstractVeniceService> createServices() {
     // Create jvm metrics object
     jvmStats = new VeniceJVMStats(metricsRepository, "VeniceJVMStats");
 
+    if (serverConfig.isSystemSchemaInitializationAtStartTimeEnabled()) {
```
Comment (lluwm) [inline services/venice-server/src/main/java/com/linkedin/venice/server/VeniceServer.java:214]: What would happen if multiple servers send concurrent `addValueSchema` to the leader controller with the same value schema contents, looks like `VeniceHelixAdmin.addValueSchema` might have an issue dealing with it?

**Comment by gaojieliu** [comment]
> > Sorry for the late follow up. Do we have a test to validate the new server behavior when enabling system store schema registration? I didn't see it from this PR.
> 
> No problem. Since VeniceServerWrapper has the following code
> 
> ```
> .put(SYSTEM_SCHEMA_INITIALIZATION_AT_START_TIME_ENABLED, true)
> ```
> 
> , when any server starts during the integration test, ControllerClientBackedSystemSchemaInitializer will be executed. If something goes wrong during the initialization, server will fail to start and the integration test will fail. Let me know if you think that more testings need to be added.

So this behavior is being tested implicitly, right?
In integration test, normally, we will spin up the Controller first and then other components, that means Venice Server won't execute the schema registration code path (very likely).
Is it possible to explicitly disable system store schema registration in Controller and exercise the Server registration path?

File: services/venice-server/src/main/java/com/linkedin/venice/server/VeniceServer.java (hunk 5 above)
Comment (shuhui-liu) [reply services/venice-server/src/main/java/com/linkedin/venice/server/VeniceServer.java:214]: It uses the pre-defined schema id to add value schema. API `addValueSchema` with value schema id is idempotent. For example, even if there are multiple calls to add value schema v15 at the same time, we will end up having only one znode /value-schema/15.

**Comment by shuhui-liu** [comment]
> > > Sorry for the late follow up. Do we have a test to validate the new server behavior when enabling system store schema registration? I didn't see it from this PR.
> > 
> > 
> > No problem. Since VeniceServerWrapper has the following code
> > ```
> > .put(SYSTEM_SCHEMA_INITIALIZATION_AT_START_TIME_ENABLED, true)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > , when any server starts during the integration test, ControllerClientBackedSystemSchemaInitializer will be executed. If something goes wrong during the initialization, server will fail to start and the integration test will fail. Let me know if you think that more testings need to be added.
> 
> So this behavior is being tested implicitly, right? In integration test, normally, we will spin up the Controller first and then other components, that means Venice Server won't execute the schema registration code path (very likely). Is it possible to explicitly disable system store schema registration in Controller and exercise the Server registration path?

I see what you mean. That's a good point. I added a new integ test VeniceServerTest#testStartServerWithSystemSchemaInitialization which disables controller meta system store initialization, lets server execute the code path, verifies that the system store is created and has the right number of schemas.

**Review by gaojieliu — APPROVED** [review]
Thanks!
